### PR TITLE
Fixed #29714 -- Allowed using ExceptionReporter subclass with AdminEmailHandler.

### PR DESCRIPTION
**Pull request by nasirhjafri**
https://code.djangoproject.com/ticket/29714
As suggested on https://code.djangoproject.com/ticket/25167

Next steps are to move getting cookies here https://github.com/django/django/blob/52e9c753659ffeab67149fcf26b95c10cf137c40/django/views/debug.py#L319 to a wrapper function so that they could be overridden.

### CODE REVIEW

**Comment by timgraham** [comment]
Look at this again (since my comment on the ticket 25167), I'm not sure if this is the right approach. If the `ExceptionReporter` only needs to be customized for `AdminEmailHandler`, it could be done with an argument to `__init__()` (similar to `include_html` and `email_backend`). This doesn't allow customizing the `ExceptionReporter` used by `technical_500_response()` though.

**Comment by nasirhjafri** [comment]
> technical_500_response

Hi @timgraham Thanks for pointing out. `technical_500_response` should use custom `ExceptionReporter` now.

File: django/views/debug.py [hunk 1]
```diff
@@ -86,7 +86,8 @@ def technical_500_response(request, exc_type, exc_value, tb, status_code=500):
     Create a technical server error response. The last three arguments are
     the values returned from sys.exc_info() and friends.
     """
-    reporter = ExceptionReporter(request, exc_type, exc_value, tb)
+    reporter_klass = get_default_exception_reporter()
+    reporter = reporter_klass(request, exc_type, exc_value, tb)
```
Comment (carltongibson) [inline django/views/debug.py]: `klass` is often used because `class` is a reserved keyword. I don't this it's necessary here. `reporter_class` would be fine.

File: tests/logging_tests/tests.py [hunk 2]
```diff
@@ -432,12 +434,22 @@ def test_disallowed_host_doesnt_crash(self):
...
+        admin_email_handler = AdminEmailHandler()
+        self.assertEqual(admin_email_handler.reporter_klass, MyExceptionReporter)
+
 
 class SettingsConfigTest(AdminScriptTestCase):
     """
     Accessing settings in a custom logging handler does not trigger
     a circular import error.
     """
+
```
Comment (carltongibson) [inline tests/logging_tests/tests.py]: Please drop these unrelated changes.

File: django/utils/log.py [hunk 3]
```diff
@@ -87,6 +87,7 @@ def __init__(self, include_html=False, email_backend=None):
         super().__init__()
         self.include_html = include_html
         self.email_backend = email_backend
+        self.reporter_klass = get_default_exception_reporter()
```
Comment (carltongibson) [inline django/utils/log.py]: Rather than adding the method call here, add a kwarg, as per the suggestion in Tim's comment. This can then be provided in the `LOGGING` config. 

See `logging.txt` in the `Handlers` section for the examples of `include_html` and `email_backend`. Similar docs are needed for the new parameter.

**Comment by nasirhjafri** [comment]
Hi, @carltongibson Thanks for reviewing. I'll do the requested changes.

**Comment by carltongibson** [comment]
Super! 🎉 Please ping me to make sure I see this 🙂

**Comment by nasirhjafri** [comment]
> Hi @nasirhjafri.
> 
> Thanks for this. Sorry for the slow review! (The _Has Patch_ flag never got updated, so it slipped the radar.)
> 
> Are you up for finishing this off? Let me know! (I think it would be a really good addition.)
> 
> I think we should look to add a kwarg to `technical_500_response()`, rather than adding a new setting here.
> 
> ```python
> def technical_500_response(..., reporter_class=ExceptionReporter):
> ```
> 
> This way users can set their own `handler500`, with a custom `reporter_class` using `functools.partial()`.
> 
> We should add an example of this in `error-reporting.txt` in the _Custom error reports_ section. (Perhaps also link there from `topics/http/views.txt` _Customizing error views_.)
> See inline comments for `AdminEmailHandler`.

@carltongibson  If the user customizes this using the above methods then there would be an inconsistency between `technical_500_response` using custom exception reporter and `AdminEmailHandler` not using that custom exception reporter.

> Rather than adding the method call here, add a kwarg, as per the suggestion in Tim's comment. This can then be provided in the `LOGGING` config.
> 
> See `logging.txt` in the `Handlers` section for the examples of `include_html` and `email_backend`. Similar docs are needed for the new parameter.

@carltongibson If we pick this setting from `LOGGING`, we can't use it in `technical_500_response`.
As Tim said 
> If the ExceptionReporter only needs to be customized for AdminEmailHandler, it could be done with an argument to __init__() (similar to include_html and email_backend). This doesn't allow customizing the ExceptionReporter used by technical_500_response() though.

To make it consistent for both `AdminEmailHandler` class and `technical_500_response` functions best solution is to add a new setting. 

I'll push other changes you requested.

**Comment by carltongibson** [comment]
> This doesn't allow customizing the ExceptionReporter used by technical_500_response() though.

Hence adding the kwarg there. (The whole point being to avoid adding a new setting.)

**Comment by nasirhjafri** [comment]
Hi @carltongibson on further investigation, I've found that custom `handler500` is being used in `DEBUG=False` context and `technical_500_response` is being used in `DEBUG=True` context. 
So we can't use `technical_500_response`  with a custom `reporter_class` using `functools.partial()`

Since we can't override `technical_500_response`, the only way to use custom `ExceptionReporter` in `technical_500_response` is to using the new setting. 

> I think we should look to add a kwarg to `technical_500_response()`, rather than adding a new setting here.
> 
> ```python
> def technical_500_response(..., reporter_class=ExceptionReporter):
> ```
> 
> This way users can set their own `handler500`, with a custom `reporter_class` using `functools.partial()`.

Need your feedback on it.

Thanks

**Comment by carltongibson** [comment]
Hi @nasirhjafri. 

> ... we can't override `technical_500_response`...

Ah, yes, you're right. Grrr. It's slightly frustrating there's no easy way to inject a custom class here. 

OK...

In that case, I think we should just solve for the `DEBUG=False` case here. (Allowing injection into `AdminEmailHandler`.) That's what the ticket was originally for, and is uncontroversial (I hope 😬)

After that they'll be some questions we can follow up with: 

1. Do we **need** to customise the Exception Reporter for `DEBUG=True`? Quite possibly we don't. 
2. If yes, can we refactor to allow injecting a custom class? 
3. If not, are we willing to add a new setting for this? (Review answer to 1.)

I'd want to go via the mailing list at that point before just adding a setting.

**Comment by nasirhjafri** [comment]
@carltongibson Can you please review the code again? I'll post the follow-up questions in the mailing list after this.

File: tests/logging_tests/logconfig.py [hunk 4]
```diff
@@ -13,3 +14,7 @@ def __init__(self):
 class MyEmailBackend(BaseEmailBackend):
     def send_messages(self, email_messages):
         pass
+
+
+class MyExceptionReporter(ExceptionReporter):
+    pass
```
Comment (felixxm) [inline tests/logging_tests/logconfig.py]: Can we customize this class to check if it really works? e.g.
```python
class MyExceptionReporter(ExceptionReporter): 
    def get_traceback_text(self):
        return 'My custom traceback ' + super().get_traceback_text()
```

**Comment by nasirhjafri** [comment]
@felixxm Pushed the requested changes, please review.

File: docs/topics/logging.txt [hunk 5]
```diff
@@ -649,6 +649,27 @@ Python logging module.
         subclass the :class:`~django.utils.log.AdminEmailHandler` class and
         override this method.
 
+
```
Comment (carltongibson) [inline docs/topics/logging.txt]: Single blank line here/

**Comment by carltongibson** [comment]
@nasirhjafri Can you then squash to a single commit? For me, this can be `Fixed #29714 -- ...`

Good stuff. Thank you for the effort! 🎉

**Comment by carltongibson** [comment]
Docs failures: 

```
/home/jenkins/workspace/docs/docs/topics/logging.txt:665: WARNING: Unexpected indentation.
/home/jenkins/workspace/docs/docs/topics/logging.txt:666: WARNING: Block quote ends without a blank line; unexpected unindent.

```

**Comment by nasirhjafri** [comment]
@carltongibson Squashed the commits and fixed the doc errors.

File: docs/topics/logging.txt [hunk 6]
```diff
@@ -649,6 +649,27 @@ Python logging module.
...
+            'mail_admins': {
+                'level': 'ERROR',
+                'class': 'django.utils.log.AdminEmailHandler',
+                'include_html': True,
+                'reporter_class': 'somepackage.error_reporter.CustomErrorReporter'
+            }
+        },
+
+    Note that ``reporter_class`` should either inherit from ``django.views.debug.ExceptionReporter``
+    or implement all methods of ``django.views.debug.ExceptionReporter``
```
Comment (felixxm) [inline docs/topics/logging.txt]: Please add `.. versionadded::`.

Comment (nasirhjafri) [reply docs/topics/logging.txt]: @felixxm I'm not sure about the version, Can you please mention which version should I add here?

Comment (felixxm) [reply docs/topics/logging.txt]: 3.0

Comment (nasirhjafri) [reply docs/topics/logging.txt]: done!

File: docs/topics/logging.txt [hunk 7]
```diff
@@ -649,6 +649,29 @@ Python logging module.
...
+
+    .. versionadded:: 3.0
```
Comment (felixxm) [inline docs/topics/logging.txt]:
```
.. versionadded:: 3.0

    The ``reporter_class`` parameter was added.
```

File: docs/topics/logging.txt (hunk 5 above)
Comment (felixxm) [reply docs/topics/logging.txt]: i.e. please chop this blank line.

File: docs/topics/logging.txt [hunk 8]
```diff
@@ -649,6 +649,29 @@ Python logging module.
...
+        'handlers': {
+            'mail_admins': {
+                'level': 'ERROR',
+                'class': 'django.utils.log.AdminEmailHandler',
+                'include_html': True,
+                'reporter_class': 'somepackage.error_reporter.CustomErrorReporter'
+            }
+        },
+
+    Note that ``reporter_class`` should either inherit from ``django.views.debug.ExceptionReporter``
```
Comment (felixxm) [inline docs/topics/logging.txt]: Wrap docs at 79 characters.

File: docs/topics/logging.txt (hunk 7 above)
Comment (nasirhjafri) [reply docs/topics/logging.txt]: done!

File: docs/topics/logging.txt (hunk 8 above)
Comment (nasirhjafri) [reply docs/topics/logging.txt]: done!

**Comment by nasirhjafri** [comment]
@carltongibson @felixxm Added a release note too!

**Comment by nasirhjafri** [comment]
@felixxm @carltongibson Please review it. Also, Is it going to be in 3.0 feature freeze?

**Comment by carltongibson** [comment]
Hi @nasirhjafri  this is on my list. So yes hopefully.

File: docs/releases/3.0.txt [hunk 9]
```diff
@@ -222,7 +222,11 @@ CSRF
 Email
 ~~~~~
 
-* ...
+* The new ``reporter_class`` parameter of
```
Comment (carltongibson) [inline docs/releases/3.0.txt]: Not sure if this wouldn't be better off in a `Logging` section...

Comment (felixxm) [reply docs/releases/3.0.txt]: Agreed.
